src/main/java/com/uid2/shared/secure/azurecc/AzurePublicKeyProvider.java

```diff
@@ -0,0 +1,77 @@
+package com.uid2.shared.secure.azurecc;
+
+import com.azure.security.attestation.AttestationClientBuilder;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.ImmutableMap;
+import com.uid2.shared.secure.AttestationException;
+
+import java.security.PublicKey;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+// MAA certs are stored as x5c(X.509 certificate chain), not supported by Google auth lib.
+// So we have to build a thin layer to fetch Azure public key.
+public class AzurePublicKeyProvider implements IPublicKeyProvider {
+
+    private final LoadingCache<String, Map<String, PublicKey>> publicKeyCache;
+
+    public AzurePublicKeyProvider() {
+        this.publicKeyCache = CacheBuilder.newBuilder()
+                .expireAfterWrite(1L, TimeUnit.HOURS)
+                .build(new CacheLoader<>() {
+                    @Override
+                    public Map<String, PublicKey> load(String maaServerBaseUrl) throws AttestationException {
+                        return loadPublicKeys(maaServerBaseUrl);
+                    }
+                });
+    }
+
+    @Override
+    public PublicKey GetPublicKey(String maaServerBaseUrl, String kid) throws AttestationException {
+        PublicKey key;
+        try {
+            key = publicKeyCache.get(maaServerBaseUrl).get(kid);
+        }
+        catch (ExecutionException e){
+            throw new AttestationException(
+                    String.format("Error fetching PublicKey from certificate location: %s, error: %s.", maaServerBaseUrl, e.getMessage())
+            );
+        }
+
+        if(key == null){
+            throw new AttestationException("Could not find PublicKey for provided keyId: " + kid);
+        }
+        return key;
+    }
+
+    // We don't want to reinvent the wheel. Leverage Azure Attestation client library to fetch certs.
+    private static Map<String, PublicKey> loadPublicKeys(String maaServerBaseUrl) throws AttestationException {
+        var attestationBuilder = new AttestationClientBuilder();
+        var client = attestationBuilder
+                .endpoint(maaServerBaseUrl)
+                .buildClient();
+
+        var signers = client.listAttestationSigners().getAttestationSigners();
+
+        ImmutableMap.Builder<String, PublicKey> keyCacheBuilder = new ImmutableMap.Builder();
+
+        for (var signer : signers){
+            var keyId = signer.getKeyId();
+            var certs = signer.getCertificates();
+            if(!certs.isEmpty()){
+                var publicKey = certs.get(0).getPublicKey();
```

Comment:
Is there any chance that a certs contains more than 1 public key, and the second key matches with the public key in JWT?

Reply:
Good question.
It's possible that there's a cert chain. But here we only care about the first level cert. We don't care about it's parent cert.
See below for example.
<img width="884" alt="image" src="https://github.com/IABTechLab/uid2-shared/assets/137876002/35abf775-4c1b-45f5-ac79-b0ef892c8fcb">

I could add some comments to elaborate in code.

Reply:
Got it, thanks for the explanation